File: src/graph/island_perimeter.rs
use crate::Solution;

const DIR: [[i32; 2]; 4] = [[0, 1], [0, -1], [1, 0], [-1, 0]];

impl Solution {
    pub fn island_perimeter(grid: Vec<Vec<i32>>) -> i32 {
        let mut ret = 0;

        for i in 0..grid.len() as i32 {
            for j in 0..grid[0].len() as i32 {
                let ii = i as usize;
                let jj = j as usize;
                // at every land, check four directions to see if there are islands near by
                if grid[ii][jj] == 1 {
                    for d in DIR {
                        let (x, y) = (i + d[0], j + d[1]);
                        if x < 0
                            || y < 0
                            || x >= grid.len() as i32
                            || y >= grid[0].len() as i32
                            || grid[x as usize][y as usize] == 0
                        {
                            ret += 1;
                        }
                    }
                }
            }
        }
        ret
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test() {
        let grid = vec![
            vec![0, 1, 0, 0],
            vec![1, 1, 1, 0],
            vec![0, 1, 0, 0],
            vec![1, 1, 0, 0],
        ];
        println!("{}", Solution::island_perimeter(grid));
    }
}

File: src/graph/island-perimeter.py
from typing import List


class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        def calc_adj(grid, i, j) -> int:
            adj = 0
            if i > 0 and grid[i - 1][j] == 1:
                adj += 1
            if i < len(grid) - 1 and grid[i + 1][j] == 1:
                adj += 1
            if j > 0 and grid[i][j - 1] == 1:
                adj += 1
            if j < len(grid[i]) - 1 and grid[i][j + 1] == 1:
                adj += 1
            return adj

        if not grid:
            return 0
        islands = 0
        adj = 0

        # count islands & adjacent times
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    islands += 1
                    adj += calc_adj(grid, i, j)
        return 4 * islands - adj


if __name__ == "__main__":
    s = Solution()
    grid = [[1]]
    print(s.islandPerimeter(grid))
